### DISCUSSION TITLE
@xstate/vue: Does it make sense to support Vue 2 and `@vue/composition/api`?

### DISCUSSION THREAD

**Discussion by fobdy** [statelyai/xstate#2193] [Q&A] [answered]
We're now using Vue 2 with `@vue/composition-api`.

But this prevents us from using the `@xstate/vue` package that imports composition API functions directly from `vue` package.

There is a [`vue-demi`](https://github.com/vueuse/vue-demi) project that addresses this type of issue by allowing you to write a code that will support Vue 2 and Vue 3 simultaneously.

I can work on it and provide a corresponding PR. But I'm not sure if it suits the author's vision and he will ever accept such an approach.

What can I do if I want to use `@xstate/vue` package in a Vue 2 world?

@davidkpiano

**Comment by davidkpiano** [comment, marked as answer]
The official recommendation for Vue 2 is to manually set up machines/services: https://xstate.js.org/docs/recipes/vue.html

To reduce maintenance burden, the `@xstate/vue` package is meant to only be supported in Vue 3, I believe (@sarahdayan might have more insight here).

From my understanding, it is not trivial to support both Vue 2 and Vue 3 in the same package without a lot of workarounds and added complexity.

If you'd like, making a Vue 2 + Composition API compatible library would be most welcomed, and I'd mention such a package in the documentation. But it should be a separate package.

**Reply by fobdy** [reply]
Thanks! Going to wait a bit until @sarahdayan may share some insights and then will mark the question as answered :)

**Reply by ChrisShank** [reply]
I just tried to port `@xstate/vue` to vue-demi... the challenge is getting the same tests to run for vue 2 and vue 3 because different versions of `jest-loader` and `@testing-library/vue` are needed depending on which version of due you are testing. otherwise it *seems* that everything still works (i.e the current tests still pass for vue 3 when using `vue-demi`). 

If the tests we refactored to test the composables in isolation (e.g . https://github.com/vueuse/vueuse/blob/main/packages/core/asyncComputed/index.test.ts) or use [vue-composition-test-utils](https://github.com/ariesjia/vue-composition-test-utils) I imagine it be pretty easy to run them for both versions of Vue.

Here is the draft https://github.com/ChrisShank/xstate/pull/3/files. Any thoughts?

**Reply by fobdy** [reply]
@ChrisShank Hm... nice thing. Going to check soon.

**Reply by Andarist** [reply]
@ChrisShank while possible this is some maintenance/config burden. Vue 3 has been released over half a year ago and it seems to me that it's an OK thing to do to only support that version. This also acts like a little push for people to upgrade to the newest version which in turn helps the Vue team a little bit - supporting multiple major versions for library maintainers is a huge pain.

**Reply by fobdy** [reply]
@Andarist The main problem that's in the Vue world things are a little bit different (rather than in the React one). Vue 3 isn't yet well supported across 3rd party libraries (such as e.g `vue-class-component` / `vue-property-decorator`). So I would say the community wasn't well prepared for migration. Official migration build is not yet released but a lot of time has passed since the release. IMHO all this leads to the fact that many commercial projects will not soon be able to switch to Vue 3.

**Reply by fobdy** [reply]
For example, in React, before breaking any kind of API, intermediate versions are released where certain things gradually become deprecated. With Vue 3, this has not happened and it is very difficult to migrate the project.

**Reply by ChrisShank** [reply]
Ya i agree with @fobdy, even though vue 3 was released a decent time ago migration hasn't been easy in my experience because the ecosystem is still upgrading and the vue core team is still working backporting new features to vue 2, ect.

but im happy to make a new package if thats the preferred solution for yall

**Comment by fobdy** [comment]
@ChrisShank It seems we may need to just make a fork and maintain a separate 3rd party vue 2 version of `@xstate/vue` without using `vue-demi`.

**Reply by davidkpiano** [reply]
I think that would be the best solution. I'll gladly link to it in the documentation.

**Reply by ChrisShank** [reply]
Totally understandable i can try to put a new package together this week! I sort of need this for the apps that i work on 😅

**Reply by ChrisShank** [reply]
@fobdy Quickly put the new package together. Need to do a little more testing, but it should be in good enough shape to publish to npm this weekend under `xstate-vue2`. https://github.com/ChrisShank/xstate-vue2

(Using vite as a dev playground and to build the package 👀)
